pkg/hubble/getters.go

```diff
@@ -0,0 +1,165 @@
+
+// Package hubble defines getter interfaces Hubble uses to access Cilium state, as
+// well as a monitor listener that sends monitor events to Hubble server.
+//
+// The getter interfaces are defined in: https://github.com/cilium/hubble/blob/master/pkg/parser/getters/getters.go
+package hubble
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	hubbleProto "github.com/cilium/hubble/api/v1/flow"
+	hubbleV1 "github.com/cilium/hubble/pkg/api/v1"
+	hubbleIPCache "github.com/cilium/hubble/pkg/ipcache"
+	hubbleEndpoint "github.com/cilium/hubble/pkg/parser/endpoint"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/identity/cache"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/service"
+)
+
+// LocalIdentityGetter implements IdentityGetter interface from Hubble.
+//  - IdentityGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L40
+type LocalIdentityGetter struct {
+	allocator *cache.CachingIdentityAllocator
+}
+
+// NewLocalIdentityGetter returns an initialized pointer to LocalIdentityGetter.
+func NewLocalIdentityGetter(allocator *cache.CachingIdentityAllocator) *LocalIdentityGetter {
+	return &LocalIdentityGetter{allocator: allocator}
+}
+
+// GetIdentity looks up identity by ID from Cilium's identity cache. Hubble uses the identity info
+// to populate source and destination labels of flows.
+func (getter *LocalIdentityGetter) GetIdentity(securityIdentity uint64) (*models.Identity, error) {
+	ident := getter.allocator.LookupIdentityByID(context.Background(), identity.NumericIdentity(securityIdentity))
+	if ident == nil {
+		return nil, fmt.Errorf("identity %d not found", securityIdentity)
+	}
+	return ident.GetModel(), nil
+}
+
+// LocalEndpointDNSGetter implements DNSGetter and EndpointGetter interfaces from Hubble.
+//  - EndpointGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L34
+//  - DNSGetter: https://github.com/cilium/hubble/blob/04ab72591faca62a305ce0715108876167182e04/pkg/parser/getters/getters.go#L27
+type LocalEndpointDNSGetter struct {
+	manager *endpointmanager.EndpointManager
+}
+
+// NewLocalEndpointDNSGetter returns an initialized pointer to LocalEndpointDNSGetter.
+func NewLocalEndpointDNSGetter(manager *endpointmanager.EndpointManager) *LocalEndpointDNSGetter {
+	return &LocalEndpointDNSGetter{manager: manager}
+}
+
+// GetEndpoint returns endpoint info for a given IP address. Hubble uses this function to populate
+// fields like namespace and pod name for local endpoints.
+func (getter *LocalEndpointDNSGetter) GetEndpoint(ip net.IP) (endpoint *hubbleV1.Endpoint, ok bool) {
+	ep := getter.manager.LookupIP(ip)
+	if ep == nil {
+		return nil, false
+	}
+	return hubbleEndpoint.ParseEndpointFromModel(ep.GetModel()), true
```

Comment:
This is extremely costly. Can't we simplify this function to be less CPU / memory intensive?

Reply:
i guess one idea would be to get rid of v1.Endpoint from hubble and use models.Endpoint instead. that way you don't need to convert from one endpoint type to another. that would be a big change on hubble side though. were you thinking about something else?

Reply:
maybe i can add another function to the endpoint getter interface that returns only the fields needed by the parser. let me see what i can do.

Reply:
@aanm  something like this?

```
	GetEndpointInfo(ip net.IP) (namespace string, name string, labels []string, ok bool)
```

Reply:
hubble side change to the endpoint getter interface: https://github.com/cilium/hubble/pull/144

Reply:
Can we define an interface that Hubble uses which Cilium can implement instead? We should be able to access the `Endpoint struct` directly and standalone Hubble can implement it with a different backing.

Reply:
I just made a similar comment on the other side: https://github.com/cilium/hubble/pull/144#issuecomment-591551852

Reply:
@aanm @glibsm How about we merge this initial inefficient version to need expand scope this PR any further. Then refactor Hubble to use an interface and then implement that interface on the Cilium side so we can return `*Endpoint` directly?

Reply:
defined an interface like @tgraf suggested, and got rid of the conversion.